fastly/fastly_alerts.go

```diff
@@ -0,0 +1,344 @@
+}
+
+// ListAlertDefinitionsInput is used as input to the ListAlertDefinitions function.
+type ListAlertDefinitionsInput struct {
+	// ServiceID filters definitions by service (optional).
+	ServiceID *string
+	// ServiceCustomerID is for Fastly Staff use only (optional).
+	ServiceCustomerID *string
+	// Name filters definitions by name substring (optional).
+	Name *string
+	// CreatedBy is for Fastly Staff use only (optional).
+	CreatedBy *string
+	// Cursor is the pagination cursor from a previous request's meta (optional).
+	Cursor *string
+	// Limit is the maximum number of items included in each response (optional).
+	Limit *int
+	// Sort is the field on which to sort definitions (optional).
+	Sort *string
+}
+
+// ListAlertDefinitions retrieves filtered, paginated alert definitions.
+func (c *Client) ListAlertDefinitions(i *ListAlertDefinitionsInput) (*AlertDefinitionsResponse, error) {
+	p := "/alerts/definitions"
+
+	ro := &RequestOptions{
+		Params: map[string]string{},
+	}
+	if i.ServiceID != nil {
+		ro.Params["service_id"] = *i.ServiceID
+	}
+	if i.ServiceCustomerID != nil {
+		ro.Params["service_customer_id"] = *i.ServiceCustomerID
+	}
+	if i.Name != nil {
+		ro.Params["name"] = *i.Name
+	}
+	if i.CreatedBy != nil {
+		ro.Params["created_by"] = *i.CreatedBy
+	}
+	if i.Cursor != nil {
+		ro.Params["cursor"] = *i.Cursor
+	}
+	if i.Limit != nil {
+		ro.Params["limit"] = strconv.Itoa(*i.Limit)
+	}
+	if i.Sort != nil {
+		ro.Params["sort"] = *i.Sort
+	}
+
+	resp, err := c.Get(p, ro)
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	var adr *AlertDefinitionsResponse
+	if err := json.NewDecoder(resp.Body).Decode(&adr); err != nil {
+		return nil, err
+	}
+
+	return adr, nil
+}
+
+// CreateAlertDefinitionInput is used as input to the CreateAlertDefinition function.
+type CreateAlertDefinitionInput struct {
+	// ServiceID is the ID of the service that the alert is monitoring (required).
+	ServiceID string `json:"service_id"`
+	// ServiceCustomerID is for Fastly staff use only (optional).
+	ServiceCustomerID string `json:"service_customer_id,omitempty"`
+	// Name is the summary text of the alert (required, limit 255).
+	Name string `json:"name"`
+	// Description is additional text included in an alert notification (optional, limit 4096).
+	Description string `json:"description"`
+	// Metric is the name of the metric being monitored for alert evaluation (required).
+	Metric string `json:"metric"`
+	// Source is the metric source (required). Options are: 'stats', 'origins', 'domains'.
+	Source string `json:"source"`
+	// Dimensions are a list of origins or domains that the alert is restricted to.
+	Dimensions map[string][]string `json:"dimensions"`
+	// EvaluationStrategy is the evaluation strategy for the alert (required).
+	EvaluationStrategy map[string]any `json:"evaluation_strategy"`
+	// IntegrationIDs are IDs of integrations that notifications will be sent to.
+	IntegrationIDs []string `json:"integration_ids"`
```

Comment:
Please ensure all _input_ structs use pointers for their field types. If the type is a slice/map/channel then those don't need to be wrapped in a pointer as they're already pointer types (e.g. the Go runtime instantiates a [runtime.slice](https://github.com/golang/go/blob/883f062fc0a097bf561030ad453fd3e300896975/src/runtime/slice.go#L15), [runtime.hmap](https://github.com/golang/go/blob/883f062fc0a097bf561030ad453fd3e300896975/src/runtime/map.go#L117) or [runtime.hchan](https://github.com/golang/go/blob/883f062fc0a097bf561030ad453fd3e300896975/src/runtime/chan.go#L33C6-L33C11) and returns a pointer to them, so we can still carry out `nil` checks).

Reply:
Done.